bofire/strategies/predictives/botorch.py

```diff
@@ -326,71 +431,58 @@ def _ask(self, candidate_count: int) -> pd.DataFrame:
 
-        if len(acqfs) > 1:
-            candidates, _ = optimize_acqf_list(
-                acq_function_list=acqfs,
-                bounds=bounds,
-                num_restarts=self.num_restarts,
-                raw_samples=self.num_raw_samples,
-                equality_constraints=get_linear_constraints(
-                    domain=self.domain, constraint=LinearEqualityConstraint  # type: ignore
-                ),
-                inequality_constraints=get_linear_constraints(
-                    domain=self.domain, constraint=LinearInequalityConstraint  # type: ignore
-                ),
-                nonlinear_inequality_constraints=nchooseks,
+        # do the global opt
+        candidates, global_acqf_val = self._optimize_acqf_continuous(
+            candidate_count=candidate_count,
+            acqfs=acqfs,
+            bounds=bounds,
+            ic_generator=ic_generator,  # type: ignore
+            ic_gen_kwargs=ic_gen_kwargs,
+            nchooseks=nchooseks,  # type: ignore
+            fixed_features=fixed_features,
+            fixed_features_list=fixed_features_list,
+        )
+
+        if (
+            self.local_search_config is not None
+            and has_local_search_region(self.domain)
+            and candidate_count == 1
+        ):
+            local_candidates, local_acqf_val = self._optimize_acqf_continuous(
+                candidate_count=candidate_count,
+                acqfs=acqfs,
+                bounds=local_bounds,
+                ic_generator=ic_generator,  # type: ignore
+                ic_gen_kwargs=ic_gen_kwargs,
+                nchooseks=nchooseks,  # type: ignore
                 fixed_features=fixed_features,
                 fixed_features_list=fixed_features_list,
-                ic_gen_kwargs=ic_gen_kwargs,
-                ic_generator=ic_generator,
-                options={"batch_limit": 5, "maxiter": 200},
             )
-        else:
-            if fixed_features_list:
-                candidates, _ = optimize_acqf_mixed(
-                    acq_function=acqfs[0],
-                    bounds=bounds,
-                    q=candidate_count,
-                    num_restarts=self.num_restarts,
-                    raw_samples=self.num_raw_samples,
-                    equality_constraints=get_linear_constraints(
-                        domain=self.domain, constraint=LinearEqualityConstraint  # type: ignore
-                    ),
-                    inequality_constraints=get_linear_constraints(
-                        domain=self.domain, constraint=LinearInequalityConstraint  # type: ignore
-                    ),
-                    nonlinear_inequality_constraints=nchooseks,
-                    fixed_features_list=fixed_features_list,
-                    ic_generator=ic_generator,
-                    ic_gen_kwargs=ic_gen_kwargs,
-                )
+            if self.local_search_config.is_local_step(
+                local_acqf_val.item(), global_acqf_val.item()
```

Comment:
`is_local_step` in the current implementation does only make use of the local optimum, right? Couldn't we save some computation time by not always compute the global optimum?

Reply:
In principle yes, and I also thought about it, but I discarded this idea, because I already have ideas regarding improvements of the method where also the global value is used, and I coded it up in a way that we can easily plug and play these new methods, by just providing it with a new `LocalSearchConfig` class and a new `is_local_step` method and it just runs without the need for touching the actual optimization code. For this reason, I traded computation time for versatility.